examples/Markers/transforms_and_units.py

```diff
@@ -0,0 +1,87 @@
+"""
+Transforms and Units
+====================
+
+This example shows how to use both the ``offsets_transform`` and ```transforms``
+parameters for markers
+"""
+
+#%%
+# Create a signal
+
+import hyperspy.api as hs
+import numpy as np
+
+signal = hs.signals.Signal1D((np.arange(100) + 1).reshape(10, 10))
+
+#%%
+# The first example shows how to draw markers which are relative to some
+# 1D signal.  This is how the EDS and EELS Lines are implemented in the
+# exspy package.
+
+
+segments = np.zeros((10, 2, 2)) # line segemnts for realative markers
+segments[:, 1, 1] = 1  # set y values end (1 means to the signal curve)
+segments[:, 0, 0] = np.arange(10).reshape(10)  # set x for line start
+segments[:, 1, 0] = np.arange(10).reshape(10)  # set x for line stop
+
+offsets = np.zeros((10,2)) # offsets for texts positions
+offsets[:, 1] = 1  # set y value for text position ((1 means to the signal curve))
+offsets[:, 0] = np.arange(10).reshape(10)  # set x for line start
+
+markers = hs.plot.markers.Lines(segments=segments,transform="relative")
+texts = hs.plot.markers.Texts(offsets=offsets,
+                              texts=["a", "b", "c", "d", "e", "f", "g", "h", "i"],
+                              sizes=10,
+                              offsets_transform="relative",
+                              shift=0.005)  # shift in axes units for some constant displacement
+signal.plot()
+signal.add_marker(markers)
+signal.add_marker(texts)
+
+#%%
+# The second example shows how to draw markers which extend to the edges of the
+# axes.  This is how the VerticalLines and HorizontalLines markers are implemented.
+
+markers = hs.plot.markers.Lines(segments=segments,
+                                transform="xaxis")
+
+signal.plot()
+signal.add_marker(markers)
+
+#%%
+# The third example shows how an offsets_transform of 'axes' can be used to annotate
+# a signal.
+
+offsets = [[.1, .5], ]  # offsets for positions
+marker1text = hs.plot.markers.Texts(offsets=np.add(offsets,[[.1,0]]),
+                      texts=["sizes=1: transform=`xaxis_scale`"],
+                      sizes=2, transform="display", offsets_transform="axes")
+marker = hs.plot.markers.Points(offsets=offsets,
+                      sizes=1, transform="xaxis_scale", offsets_transform="axes")
```

Comment:
Yea... So I'm pretty sure this is something that we would have to fix upstream in Matplotlib as this is consistent with (almost) all of the Matplotlib code I've seen. 

If you set the `transform` to either "axis" and "data" along with a `offset_transform` of "axis" and "data"  somethings will behave a little weird.  I don't exactly understand it but I think it is a result of both of those transforms scaling around the lower left corner of the plot and as a result they interact in very strange ways.  If you look at the ax.transData the Transformation matrix includes a shift as well.  

Currently the `xaxis_scale` and `yaxis_scale` won't track when you scale some plot as the transformation just looks at the ax.view_limits when the plot is first initialized.  You would need some affine matrix to scale the size which updates when the view_limits are changed. I've tried to get that to work but I don't think it really does...

Reply:
At some point we might have to decide that this isn't worth it to fix and call things good (or much better than they previously were). We can always make the change upstream to `matplotlib` and add proper scaling later.  

I can't quite find where the zoom/pan features of `matplotlib` are located which makes this a little bit more difficult to fix.

Reply:
You are right, as far as I understand, the issue comes from matplotlib collections implementation. Here is a summary of what I understand in order to try to find a way forward:
- `offsets_transform` is working fine for all collection type
- only `matplotlib.collections.EllipseCollection` correctly supports using the `data` coordinate system for the "size" type argument and this can be done using the `units` argument. This is the feature that we need for all markers. For example, `matplotlib.collections.CircleCollection` or any collection that inherit from `matplotlib.collections._CollectionWithSizes` doesn't support it correctly: `transform` argument seems to be ignored and using `set_transform(as.transData)` doesn't work correctly
    ```python
    import matplotlib.pyplot as plt
    import matplotlib.collections as mc
    import numpy as np
    
    sizes = np.arange(10)
    offsets = np.stack([sizes]*2).T
    
    fig, ax = plt.subplots()
    cc = mc.CircleCollection(sizes=sizes,
                             offsets=offsets,
                             offset_transform=ax.transData,
                             )
    cc.set_transform(ax.transData)
    ax.add_collection(cc)
    
    ax.set_aspect('equal')
    ax.set_xlim(0, 10)
    ax.set_ylim(0, 10)
    ```

Defining the "size" type argument in "data" coordinate system is the behaviour of the markers before this PR and we need to keep this behaviour. In this PR, this is an issue for:
- `Circles`, `Points` since they used `CircleCollection`, this can be sorted by using `EllipseCollection` instead
- `Squares`: either remove it or use the `RectangleCollection` of this PR, which uses the same approach as `EllipseCollection`
- `Polygons: we can just removed, as this is a new feature and it simply wrap the `matploltib.collections.PolyCollection`

One thing that we need to keep in mind is how to load markers which has been created with previous version of hyperspy >=2.0, typically any kwargs that we introduce now, we want to avoid having to change in future release.

I would be inclined to use the following approach:
- Using the `_CollectionWithWidthHeightAngle` should be safe from a future proofing perspective, as it is still fairly well aligned with matplotlib and rely on matplotlib private API. It should be possible to upstream this in matplotlib and at the same time sorted out some of the issues discussed above.
- Keep the `offset_transform` as it is currently
- remove the `transform` argument and we don't do anything with it. If users want to do something with it, they would need to use `set_transform` and do it at their own risk. Once the various issue discussed above are sorted in matplotlib, we can think to document this behaviour. What is the use case for xaxis_scale` and `yaxis_scale`? As implemented in this PR, the `xaxis_scale` and `yaxis_scale` are incorrect when zooming in the figure.

@CSSFrancis, what do you think?

Reply:
> * Using the `_CollectionWithWidthHeightAngle` should be safe from a future proofing perspective, as it is still fairly well aligned with matplotlib and rely on matplotlib private API. It should be possible to upstream this in matplotlib and at the same time sorted out some of the issues discussed above.

I really like this idea for the Squares and Circles! This would remove the need to `xaxis_scale` and `yaxis_scale` as well as the `transform` argument. All of the makers from `CollectionWithWidthHeightAngle` would then have a "units" parameter which would replace the `transform` argument.  

> * Keep the `offset_transform` as it is currently
> * ~remove the `transform` argument and we don't do anything with it. If users want to do something with it, they would need to use `set_transform` and do it at their own risk. Once the various issue discussed above are sorted in matplotlib, we can think to document this behaviour.~ EDIT: not possible, because `verts` `segments` uses this transform...

Yea the `verts` and `segments` require the `transform` to be set. 

> * What is the use case for `xaxis_scale` and `yaxis_scale`? As implemented in this PR, the `xaxis_scale` and `yaxis_scale` are incorrect when zooming in the figure.

Yea let's remove them in favor of the `axis` argument.

> 
> @CSSFrancis, what do you think?